crates/ruma-api/src/lib.rs

```diff
@@ -591,3 +595,82 @@ impl MatrixVersion {
         }
     }
 }
+
+/// A specifier for what path variant of an endpoint to request.
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum EndpointPath {
+    /// Select an appropriate path version for a particular matrix version.
+    ///
+    /// Note: When using
+    /// [`try_into_http_request`](OutgoingRequest::try_into_http_request) with this variant, it can
+    /// emit additional errors;
+    /// - [`NoUnstablePath`](IntoHttpError::NoUnstablePath), if the matrix version is too new, and
+    ///   the endpoint has no defined `unstable` path (for maintenance or legacy reasons.)
+    /// - [`EndpointRemoved`](IntoHttpError::EndpointRemoved), if the matrix version is too old,
+    ///   and the endpoint is known to be removed in the same or a newer version.
+    ForMatrixVersion(MatrixVersion),
+
+    /// Select any path, preferring stable.
+    ///
+    /// Note: This does not emit any errors like
+    /// [`ForMatrixVersion`](EndpointPath::ForMatrixVersion) does, if the server's lowest matrix
+    /// version has removed the endpoint, or it has not yet stabilized the endpoint, this may
+    /// result in runtime HTTP errors, it is highly recommended to fetch `/versions`, extract
+    /// Ruma-known matrix versions, and pass the preferred version through `ForMatrixVersion`.
+    PreferStable,
```

Comment:
I'm wondering whether we should leave this out for now and only have the single variant `ForMatrixVersion` for a start. You were also saying it's not that useful currently, right?

Reply:
It is useful, but only for clients which *find* it useful, i said that it's a trade-off for the library consumer to make, to choose this variant and opt out of the stricter checking ForMatrixVersion does, at the benefit of a "simpler" integration with ruma.

Reply:
Maybe, but if a server suddenly moves on to 1.1, while this still says 1.0, then those endpoints would still use r0 while all the capability to use v3 stable endpoints exist. This would also then become a wart in the codebase, so I think this variant is still useful on that from.

With this, at the very least the application would be using modern stable paths, which have a greater chance to be accepted by a server than unstable or r0 path, at one such point in the future.